_posts/2020-02-25-password-encoding-spring-security.md

```diff
@@ -0,0 +1,535 @@
+        this.jdbcUserDetailsService = jdbcUserDetailsService;
+     }
+     
+
+     @Autowired
+     public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
+        auth.authenticationProvider(daoAuthenticationProvider());
+     }
+    
+     public AuthenticationProvider daoAuthenticationProvider() {
+        DaoAuthenticationProvider daoAuthenticationProvider = new DaoAuthenticationProvider();
+        daoAuthenticationProvider.setPasswordEncoder(passwordEncoder());
+        daoAuthenticationProvider.setUserDetailsService(this.jdbcUserDetailsService);
+        return daoAuthenticationProvider;
+     }
+    
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+       return new BCryptPasswordEncoder();
+    }
+}
+````
+We created a `DaoAuthenticationProvider` and set the password encoder `BCryptPasswordEncoder` to the authentication provider.
+That's it for the password encoding. This `BCryptPasswordEncoder` will be automatically used by `AuthenticationProvider`
+for password encoding and password matching. Now we have to make one step more to complete the configuration. We 
+set the `JdbcUserDetailsService` service to the `DaoAuthenticationProvider`. After that `DaoAuthenticationProvider` can 
+get the user data to execute the authentication. Spring Security takes care of the rest. If a client sends a 
+HTTP request with basic authentication header, Spring Security will read this header, load data for the user,
+who wants to authenticate and matches the password using `BCryptPasswordEncoder`. If the password matches the request will be passed
+through, if not the server will send a response with status code 401.
+
+To add a user to the system, let's implement the API for registration.
+````java
+@RestController
+@Transactional
+public class UserResources {
+
+    private final UserRepository userRepository;
+    private final PasswordEncoder passwordEncoder;
+
+    public UserResources(UserRepository userRepository, PasswordEncoder passwordEncoder) {
+        this.userRepository = userRepository;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    @PostMapping("/registration")
+    @ResponseStatus(code = HttpStatus.CREATED)
+    public void register(@RequestBody UserCredentialsDto userCredentialsDto) {
+        UserCredentials user =
+                UserCredentials.builder()
+                        .enabled(true)
+                        .username(userCredentialsDto.getUsername())
+                        .password(passwordEncoder.encode(userCredentialsDto.getPassword()))
+                        .roles(Set.of("USER"))
+                        .build();
+        userRepository.save(user);
+    }
+}
+```` 
+As we defined in Spring Security rules, the access to `/regestration` API is open for everybody. We use the `PasswordEncoder`, that is defined in
+Spring Security configuration to encode the password. In this example, the passwords are encoded with `bcrypt` algorithm. 
```

Comment:
You can't see that from the code and that's why I add this sentence as an explanation.
I changed it to "In this example, the passwords are encoded with the `bcrypt` algorithm because we set the `PasswordEncoder`
as the password encoder in the configuration."